persistence/sql/persister_consent.go

```diff
@@ -214,146 +211,186 @@ func (p *Persister) GetConsentRequest(ctx context.Context, challenge string) (*c
 	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.HandleConsentRequest")
 	defer span.End()
 
-	f := &flow.Flow{}
-
-	if err := sqlcon.HandleError(p.QueryWithNetwork(ctx).Where("consent_challenge_id = ?", r.ID).First(f)); errors.Is(err, sqlcon.ErrNoRows) {
-		return nil, err
+	if f == nil {
+		return nil, errorsx.WithStack(fosite.ErrInvalidRequest.WithDebug("Flow was nil"))
 	}
-
+	if f.NID != p.NetworkID(ctx) {
+		return nil, errorsx.WithStack(x.ErrNotFound)
+	}
+	// Restore the short challenge ID, which was previously sent to the encoded flow.
+	r.ID = f.ConsentChallengeID.String()
 	if err := f.HandleConsentRequest(r); err != nil {
 		return nil, errorsx.WithStack(err)
 	}
 
-	_, err := p.UpdateWithNetwork(ctx, f)
-	if err != nil {
-		return nil, sqlcon.HandleError(err)
-	}
-
-	return p.GetConsentRequest(ctx, r.ID)
+	return f.GetConsentRequest(), nil
 }
 
-func (p *Persister) VerifyAndInvalidateConsentRequest(ctx context.Context, verifier string) (*consent.AcceptOAuth2ConsentRequest, error) {
+func (p *Persister) VerifyAndInvalidateConsentRequest(ctx context.Context, f *flow.Flow, verifier string) (*flow.AcceptOAuth2ConsentRequest, error) {
 	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.VerifyAndInvalidateConsentRequest")
 	defer span.End()
 
-	var r consent.AcceptOAuth2ConsentRequest
-	return &r, p.transaction(ctx, func(ctx context.Context, c *pop.Connection) error {
-		var f flow.Flow
-		if err := p.QueryWithNetwork(ctx).Where("consent_verifier = ?", verifier).First(&f); err != nil {
-			return sqlcon.HandleError(err)
-		}
+	if f == nil {
+		return nil, errorsx.WithStack(fosite.ErrInvalidRequest.WithDebug("Flow was nil"))
+	}
+	if f.NID != p.NetworkID(ctx) {
+		return nil, errorsx.WithStack(sqlcon.ErrNoRows)
+	}
 
-		if err := f.InvalidateConsentRequest(); err != nil {
-			return errorsx.WithStack(fosite.ErrInvalidRequest.WithDebug(err.Error()))
-		}
+	updatedFlow, err := flowctx.Decode[flow.Flow](ctx, p.r.FlowCipher(), verifier, flowctx.AsConsentVerifier)
+	if err != nil {
+		return nil, errorsx.WithStack(fosite.ErrAccessDenied.WithHint("The consent verifier has already been used, has not been granted, or is invalid."))
+	}
+	if updatedFlow.ID != f.ID {
+		return nil, errorsx.WithStack(fosite.ErrInvalidRequest.WithDebug("Consent verifier does not match login request."))
+	}
+	if updatedFlow.NID != p.NetworkID(ctx) {
+		return nil, errorsx.WithStack(sqlcon.ErrNoRows)
+	}
 
-		r = *f.GetHandledConsentRequest()
-		_, err := p.UpdateWithNetwork(ctx, &f)
-		return err
-	})
+	// Update flow from login request, but keep requested at.
+	updatedFlow.NID = f.NID
+	updatedFlow.ConsentCSRF = f.ConsentCSRF
+	updatedFlow.ConsentVerifier = f.ConsentVerifier
+	*f = *updatedFlow
+
+	if err = f.InvalidateConsentRequest(); err != nil {
+		return nil, errorsx.WithStack(fosite.ErrInvalidRequest.WithDebug(err.Error()))
+	}
+
+	f.ConsentChallengeID = sqlxx.NullString(uuid.Must(uuid.NewV4()).String())
```

Comment:
Why set this here?

Reply:
We set the consent challenge ID to a new UUID that we can use as a foreign key in the database without encoding the whole flow.